pkg/agent/client.go

```diff
@@ -1,54 +1,134 @@
+// Copyright 2020 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
 package agent
 
 import (
 	"fmt"
 	"io/ioutil"
 	"net"
 	"os"
+	"sync"
 
+	"k8s.io/apiserver/pkg/server/dynamiccertificates"
+	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/clientcmd"
 	"k8s.io/component-base/config"
 	"k8s.io/klog"
 
+	cert "github.com/vmware-tanzu/antrea/pkg/apiserver/certificate"
 	"github.com/vmware-tanzu/antrea/pkg/client/clientset/versioned"
 )
 
-// CreateAntreaClient creates an Antrea client from the given config.
-func CreateAntreaClient(config config.ClientConnectionConfiguration) (versioned.Interface, error) {
+// AntreaClientProvider provides a method to get Antrea client.
+type AntreaClientProvider interface {
+	GetAntreaClient() (versioned.Interface, error)
+}
+
+// antreaClientProvider provides a AntreaClientProvider that can dynamically react to ConfigMap changes.
+type antreaClientProvider struct {
+	config config.ClientConnectionConfiguration
+	// mutex protects client.
+	mutex sync.RWMutex
+	// client is the Antrea client that will be returned. It will be updated when caBundle is updated.
+	client versioned.Interface
+	// caContentProvider provides the very latest content of the ca bundle.
+	caContentProvider *dynamiccertificates.ConfigMapCAController
+}
+
+var _ dynamiccertificates.Listener = &antreaClientProvider{}
+
+func NewAntreaClientProvider(config config.ClientConnectionConfiguration, kubeClient kubernetes.Interface) *antreaClientProvider {
+	antreaCAProvider, _ := dynamiccertificates.NewDynamicCAFromConfigMapController("antrea-ca", cert.CAConfigMapNamespace, cert.CAConfigMapName, cert.CAConfigMapKey, kubeClient)
```

Comment:
could you add a comment for what happens if the key does not exist yet? I think it is handled gracefully by the apiserver (will keep retrying), but a comment may help here.